### PR TITLE
Add support for SLF4J binding via SPI in SLF4J 1.8 onwards

### PR DESCRIPTION
**Pull request by f4lco**
### Description

Linked issue: #119

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Code style follows the tinylog standard
- [x] All classes and methods have Javadoc
- [x] Changes are covered by JUnit tests including corner cases, errors, and exception handling
- [x] Maven build works including compiling, tests, and checks (`mvn verify`)
- [x] Changes are committed by a verified email address that is assigned to the GitHub account (https://github.com/settings/emails)

### Documentation

Additions or amendments for the [public documentation](https://github.com/pmwmedia/tinylog/wiki/Documentation):

```markdown
This patch allows to use Tinylog with SLF4J 1.8 and onwards, which requires locating SLF4J bindings via the service loader mechanism.
```

### Agreements

- [x] I agree that my changes will be published under the terms of the [Apache License 2.0](https://github.com/pmwmedia/tinylog/blob/v2.0/license.txt)
- [x] I agree that my GitHub user name will be published in the release notes

### Notes

- I didn't really see a good use-case for a unit test. It's really plain, devoid of any particular logic. Please feel free to give examples of testable properties if you think that's not the case.
- Also suspected in #119, using the SPI is **not** possible from SLF4J 1.7. Only SLF4J 1.8 and onwards contains the required interface.
- I'm not sure about the deprecation of the former method. At some point, we could likely remove the static binder classes `org.slf4j.impl.{StaticLoggerBinder,StaticMarkerBinder,StaticMDCBinder}` from Tinylog.
- It's me.... it's just been a while since #70 😄

### CODE REVIEW

File: pom.xml [hunk 1]
```diff
@@ -87,7 +87,7 @@
 		<mockk.version>1.11.0</mockk.version>
 		<powermock.version>2.0.9</powermock.version>
 		<scala.version>2.12.13</scala.version>
-		<slf4j.version>1.7.30</slf4j.version>
+		<slf4j.version>1.8.0-beta4</slf4j.version>
```
Comment (pmwmedia) [inline pom.xml]: I think a stable version of SLF4J 1.8.0 will be never released in favor of 2.0. Could you update the version to the latest 2.0 (alpha) version?

File: slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2018 Martin Winandy
```
Comment (pmwmedia) [inline slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java]: Please update the year to 2021. You can also replace my name by your's if you want.

File: slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+/**
+ * SPI implementation for SLF4j 1.8+. Replacement mechanism for {@link org.slf4j.impl.StaticLoggerBinder}.
+ */
+public class TinylogSlf4jServiceProvider implements SLF4JServiceProvider {
+
+	/**
+	 * SLF4J API 1.8 and newer is supported.
+	 */
+	public static final String REQUESTED_API_VERSION = "1.8";
```
Comment (pmwmedia) [inline slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java:30]: Maybe we have to replace the API version to 2.0, if SLF4J 1.8 and 2.0 are not compatible.

**Review by pmwmedia — CHANGES_REQUESTED** [review]

**Comment by pmwmedia** [comment]
> * If you don't think JitPack is warranted, I'd also drop it. I just needed it to immediately test the change in my current project. I might keep the file to my fork only, or just might use the next release of Tinylog. But I see some value in being able to run with individual commits just by adding the repository definition and using dependency coordinates such as `com.github.f4lco.tinylog:slf4j-tinylog:f93accf1c581682298246662fe5e31485d5b1c4a`.

I didn't know JitPack before, but I actually like the idea. However, I would prefer the integration by a separate pull request.

> * I didn't really see a good use-case for a unit test. It's really plain, devoid of any particular logic. Please feel free to give examples of testable properties if you think that's not the case.

A full integration test case would be great to ensure that SLF4J's `LoggerFactory` provides loggers, which are an instance of `TinylogLogger` when using SLF4J 2.0.

> * Also suspected in #119, using the SPI is **not** possible from SLF4J 1.7. Only SLF4J 1.8 and onwards contains the required interface.

Older SLF4J versions should just ignore the new class. I don't expect any problems when using SLF4J 1.7. However, please correct me if I'm wrong.

> * I'm not sure about the deprecation of the former method. At some point, we could likely remove the static binder classes `org.slf4j.impl.{StaticLoggerBinder,StaticMarkerBinder,StaticMDCBinder}` from Tinylog.

SLF4J 1.7 is still very popular and there is sill no stable release of SLF4J 1.8 nor 2.0. I assume we will have to live quite long with the static binder classes.

> * It's me.... it's just been a while since #70 😄

Welcome back :) I'm glad to receive a pull request again from you!

File: pom.xml (hunk 1 above)
Comment (f4lco) [reply pom.xml]: Sure thing 👍

File: slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java (hunk 2 above)
Comment (f4lco) [reply slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java]: Will do. That's not really important to me, but thanks for suggesting, that's kind 😃

File: slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java (hunk 3 above)
Comment (f4lco) [reply slf4j-tinylog/src/main/java/org/tinylog/slf4j/TinylogSlf4jServiceProvider.java:30]: At this stage, it does not look like SLF4J has already defined compatible versions, or updated the compatibility list:

```
    /**
     * It is LoggerFactory's responsibility to track version changes and manage
     * the compatibility list.
     * <p>
     * <p>
     * It is assumed that all versions in the 1.6 are mutually compatible.
     */
    static private final String[] API_COMPATIBILITY_LIST = new String[] { "1.8", "1.7" };
```

https://github.com/qos-ch/slf4j/blob/e2c29c81df6fb405a3c5a6dfd1cf28ba26a7c1d6/slf4j-api/src/main/java/org/slf4j/LoggerFactory.java#L110-L117

So in theory we could also report 1.7, which is pretty confusing, because that would mean `TinylogSlf4jServiceProvider` requests an API level in which it could not have existed, because `SLF4JServiceProvider` did not exist. That why I believe 1.8 is the correct number to report right now.

**Comment by f4lco** [comment]
> I didn't know JitPack before, but I actually like the idea. However, I would prefer the integration by a separate pull request.

I will arrange for that.

> A full integration test case would be great to ensure that SLF4J's LoggerFactory provides loggers, which are an instance of TinylogLogger when using SLF4J 2.0.

I was a bit surprised something like that did not exist before for the static logger binder. No worries, I'll add one.

> Older SLF4J versions should just ignore the new class. I don't expect any problems when using SLF4J 1.7. However, please correct me if I'm wrong.

That's also my belief. 

> SLF4J 1.7 is still very popular and there is sill no stable release of SLF4J 1.8 nor 2.0. I assume we will have to live quite long with the static binder classes.

OK, then maybe that's in for Tinylog 3 😉 

> Welcome back :) I'm glad to receive a pull request again from you!

Thanks 😄  Please enjoy my reworked PR, your feedback was very valuable.

**Comment by pmwmedia** [comment]
The integration tests looks very good but SLF4J seems not using the new TinylogSlf4jServiceProvider for resolving the logger factory nor the MDC adapter. According to the code coverage report, none of the tests triggers any call of the TinylogSlf4jServiceProvider. 

@f4lco Do you understand the reason?

**Comment by f4lco** [comment]
> The integration tests looks very good but SLF4J seems not using the new TinylogSlf4jServiceProvider for resolving the logger factory nor the MDC adapter. According to the code coverage report, none of the tests triggers any call of the TinylogSlf4jServiceProvider.
>
> @f4lco Do you understand the reason?

No. I do not understand why @codecov did not provide us with an updated report within this PR. During test development I set breakpoints in the respective methods of `TinylogSlf4jServiceProvider` and the test definitely hit them. After a bit of digging, the report under https://codecov.io/gh/tinylog-org/tinylog/commit/4ce0fb5991d05642809f9257c37ac87197474734/ reflects that 🤷‍♂️

**Comment by pmwmedia** [comment]
> No. I do not understand why @codecov did not provide us with an updated report within this PR. During test development I set breakpoints in the respective methods of `TinylogSlf4jServiceProvider` and the test definitely hit them. After a bit of digging, the report under https://codecov.io/gh/tinylog-org/tinylog/commit/4ce0fb5991d05642809f9257c37ac87197474734/ reflects that 🤷‍♂️

I'm relived that the provider does work as expected :) @codecov did even update the report post after your commits but hasn't shown the correct code coverage for any reason.

Just a last wish before merging the pull request: Could you add a test for `MarkerFactory.getIMarkerFactory()`? tinylog converts markers into tags what requires an actual marker factory.

**Comment by f4lco** [comment]
> Just a last wish before merging the pull request: Could you add a test for MarkerFactory.getIMarkerFactory()? tinylog converts markers into tags what requires an actual marker factory.

Done 👍

**Comment by pmwmedia** [comment]
Thank you, I'm happy to merge your pull request :)
